app/src/main/java/org/simple/clinic/location/LocationRepository.kt

```diff
@@ -1,13 +1,84 @@
 package org.simple.clinic.location
 
+import android.annotation.SuppressLint
+import android.app.Application
+import android.location.Location
+import com.google.android.gms.location.FusedLocationProviderClient
+import com.google.android.gms.location.LocationAvailability
+import com.google.android.gms.location.LocationCallback
+import com.google.android.gms.location.LocationRequest
+import com.google.android.gms.location.LocationResult
+import com.google.android.gms.location.LocationServices
 import io.reactivex.Observable
 import org.threeten.bp.Duration
 import javax.inject.Inject
 
-class LocationRepository @Inject constructor() {
+typealias LocationProvider = FusedLocationProviderClient
+
+@SuppressLint("MissingPermission")
+class LocationRepository @Inject constructor(private val appContext: Application) {
+
+  data class LocationStatus(val isUsable: Boolean)
 
   fun streamUserLocation(updateInterval: Duration): Observable<LocationUpdate> {
-    // TODO.
-    return Observable.just(LocationUpdate.TurnedOff)
+    val locationProvider = LocationServices.getFusedLocationProviderClient(appContext)
+    val request = locationRequest(updateInterval)
+
+    val availableStream = locationUpdates(locationProvider, request)
+        .map { LocationUpdate.Available(it) }
+
+    val unavailableStream = locationStatusChanges(locationProvider, request)
+        .filter { it.isUsable.not() }
+        .map { LocationUpdate.Unavailable }
+
+    return Observable.merge(availableStream, unavailableStream)
+  }
+
+  private fun locationRequest(updateInterval: Duration): LocationRequest {
+    return LocationRequest().apply {
+      priority = LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY
+      interval = updateInterval.toMillis()
+    }
+  }
+
+  private fun locationUpdates(provider: LocationProvider, request: LocationRequest): Observable<Coordinates> {
+    val locationChanges = Observable.create<LocationResult> { emitter ->
+      val locationCallback = object : LocationCallback() {
+        override fun onLocationResult(result: LocationResult) {
+          emitter.onNext(result)
+        }
+      }
+
+      emitter.setCancellable { provider.removeLocationUpdates(locationCallback) }
+      provider.requestLocationUpdates(request, locationCallback, appContext.mainLooper)
```

Comment:
Because of this, all emissions from here will be emitted on the main thread. This is not evident to users of this class. Is there something we can do to make this more evident?

Reply:
Discussed in person: I'm doing this in an upcoming PR